pkg/networkservice/common/heal/eventloop.go

```diff
@@ -78,55 +76,113 @@ func newEventLoop(ctx context.Context, cc grpc.ClientConnInterface, conn *networ
 		return nil, errors.WithStack(err)
 	}
 
+	logger := log.FromContext(ctx).WithField("heal", "eventLoop")
 	cev := &eventLoop{
-		eventLoopCtx:    eventLoopCtx,
-		chainCtx:        ctx,
-		conn:            conn,
-		eventFactory:    ev,
-		client:          newClientFilter(client, conn, log.FromContext(ctx)),
-		livelinessCheck: livelinessCheck,
+		heal:         heal,
+		eventLoopCtx: eventLoopCtx,
+		chainCtx:     ctx,
+		conn:         conn,
+		eventFactory: ev,
+		client:       newClientFilter(client, conn, logger),
+		logger:       logger,
 	}
 
 	// Start the eventLoop
 	go cev.eventLoop()
 	return eventLoopCancel, nil
 }
 
-func (cev *eventLoop) eventLoop() {
-	/* Receive monitor events */
-	for {
-		eventIn, err := cev.client.Recv()
-		if cev.eventLoopCtx.Err() != nil {
-			return
-		}
+func (cev *eventLoop) waitCtrlPlaneEvent() <-chan struct{} {
+	res := make(chan struct{}, 1)
+
+	go func() {
+		defer close(res)
+
+		for {
+			eventIn, err := cev.client.Recv()
+
+			if cev.chainCtx.Err() != nil || cev.eventLoopCtx.Err() != nil {
+				res <- struct{}{}
+				return
+			}
 
-		if !cev.livelinessCheck(cev.conn) {
 			// Handle error
 			if err != nil {
 				s, _ := status.FromError(err)
 				// This condition means, that the client closed the connection. Stop healing
 				if s.Code() == codes.Canceled {
-					return
+					res <- struct{}{}
 				}
 				// Otherwise - Start healing
-				break
+				return
 			}
 
 			// Handle event. Start healing
 			if eventIn.GetConnections()[cev.conn.GetId()].GetState() == networkservice.State_DOWN {
-				break
+				return
 			}
 		}
+	}()
+
+	return res
+}
+
+func (cev *eventLoop) eventLoop() {
+	reselect := false
+
+	ctrlPlaneCh := cev.waitCtrlPlaneEvent()
+
+	livenessCheckerCtx, livenessCheckerCancel := context.WithCancel(context.Background())
+	dataPlaneCh := make(chan struct{}, 1)
+	defer livenessCheckerCancel()
+	if cev.heal.livenessChecker != nil {
+		go func() {
+			cev.heal.livenessChecker(livenessCheckerCtx, cev.conn)
+			dataPlaneCh <- struct{}{}
+		}()
+	} else {
+		// Since we don't know about data path status - always use reselect
+		reselect = true
 	}
+
+	select {
+	case _, ok := <-ctrlPlaneCh:
+		cev.logger.Warnf("Control plane is down")
+		if ok {
+			// Connection closed
+			return
+		}
+		// Start healing
+	case <-dataPlaneCh:
+		cev.logger.Warnf("Data plane is down")
+		reselect = true
+		// Start healing
+	case <-cev.chainCtx.Done():
+		return
+	case <-cev.eventLoopCtx.Done():
+		return
+	}
+
 	/* Attempts to heal the connection */
-	afterTicker := clock.FromContext(cev.eventLoopCtx).Ticker(attemptAfter)
+	afterTicker := clock.FromContext(cev.eventLoopCtx).Ticker(cev.heal.attemptAfter)
 	defer afterTicker.Stop()
 	for {
 		select {
 		case <-cev.chainCtx.Done():
 			return
+		case <-dataPlaneCh:
+			cev.logger.Warnf("Data plane is down")
+			reselect = true
 		case <-afterTicker.C():
```

Comment:
What if request try was too long and the ticker channel is not empty?

Reply:
Added `cev.chainCtx.Err() != nil` check. Also replaced `case <-dataPlaneCh` with `dataPlaneCtx.Err() != nil` check.